### DISCUSSION TITLE
Code generator plugin, how to open new sketch window and populate with generated code?

### DISCUSSION THREAD

**Discussion by PaulStoffregen** [dankeboy36/vscode-arduino-api#16] [General]
I'm trying to recreate a [Java plugin I wrote for the old Arduino IDE](https://github.com/PaulStoffregen/secure_plugin/tree/main), which helps the user to create an encryption key, and can automatically generates a few sketches with their key embedded in the sketch code.  Learning typescript as I go, but happy to say my [effort so far](https://github.com/PaulStoffregen/secure_plugin_vscode) has the first key generation part working in Arduino IDE 2.3.3 thanks to this API and quite a lot of studying [Earle F. Philhower's arduino-littlefs-upload plugin](https://github.com/earlephilhower/arduino-littlefs-upload).

But now I'm stuck on the part of creating a new sketch and populating it with generated code.  Hoping you can give me a hint on how to accomplish this?

Something I tried which didn't work was trying to create a [Virtual Document using vscode.TextDocumentContentProvider](https://code.visualstudio.com/api/extension-guides/virtual-documents).  I did manage to get generated code to appear as a new tab in the current sketch...

![image](https://github.com/user-attachments/assets/a88f243d-bab0-4ee2-a6c3-210ebcbbc15a)

Sadly, Arduino IDE seems to completely ignore this tab.

In the old Java IDE, my plugin just created a temporary folder, wrote the code into files, and then called Base.INSTANCE.handleOpen(newFile).  Really hoping there's a way to do something similar with the new IDE.  Maybe it's "obvious" and I've just been missing something?

**Comment by dankeboy36** [comment]
> But now I'm stuck on the part of creating a new sketch and populating it with generated code.

You can use [fs](https://nodejs.org/api/fs.html) from Node.js or one of your tools (and spawn a process) to generate files and folders.

> Something I tried which didn't work was trying to create a [Virtual Document using vscode.TextDocumentContentProvider](https://code.visualstudio.com/api/extension-guides/virtual-documents). I did manage to get generated code to appear as a new tab in the current sketch...

What do you want to do with the generated files? Where should the extension generate them?

I created a short example of how to generate a sketch from user input into the user's sketchbook: https://github.com/PaulStoffregen/secure_plugin_vscode/commit/43e46aa2365022658544187c4cfacc51100e4f1d.

In action:


https://github.com/user-attachments/assets/fe368c27-5c87-4b51-bac9-3e6168817dac


> Maybe it's "obvious" and I've just been missing something?

It's not trivial; I also had to look into Theia's code to see how to achieve it.

> Base.INSTANCE.handleOpen(newFile)

You can use built-in VS Code commands to open a new window: https://code.visualstudio.com/api/references/commands. My example includes a few more links.

----

I adjusted (https://github.com/PaulStoffregen/secure_plugin_vscode/commit/9187b7777680a0a2d84f70a432be78912694479d) the `vscode-arduino-api` version to the IDE2: https://github.com/arduino/arduino-ide/blob/77136687d374dce1660191059427f1941f4cef16/electron-app/package.json#L199. Although, I have released a new version with multi-sketch support, the Arduino IDE still have to support it.


I hope this helps, thanks for using the API, and if you have any questions, feel free to ask.

**Comment by dankeboy36** [comment]
From https://github.com/PaulStoffregen/secure_plugin_vscode/blob/0170248881501a65808d483d67f3fccfe82bc702/src/extension.ts#L59-L65:

> 	/*  not working, but why???
>	acontext.onDidChangeSketch((event) => {

Arduino IDE supports a single sketch per workspace; once the "current" sketch is loaded in IDE2, it won't unload or change. Hence, this value changes only once per sketch window from `undefined` to the sketch folder path.

If the API and your extension run in another environment that supports multiple sketches per workspace, for example, VS Code, this value could change more than once per workspace.

**Comment by PaulStoffregen** [comment]
Wow, thanks!  Tried `vscode.commands.executeCommand('vscode.openFolder'` just now, and it works!!!!

**Comment by PaulStoffregen** [comment]
![image](https://github.com/user-attachments/assets/f2d2f890-6804-44df-87dc-0fe807fde3e1)

**Comment by PaulStoffregen** [comment]
An exciting moment on my 4th day of learning Typescript, got it to actually fill in the generated code!

![image](https://github.com/user-attachments/assets/6031c071-6060-494a-8846-23c3916d0882)

**Comment by PaulStoffregen** [comment]
Quick novice question... I brought in your [changes to package.json and package-lock.json](https://github.com/PaulStoffregen/secure_plugin_vscode/commit/9187b7777680a0a2d84f70a432be78912694479d).  Everything seems fine when I build with "npm run compile".  But if I try running "vsce package", I'm getting this error:

```
Executing prepublish script 'npm run vscode:prepublish'...

> teensysecurity@0.0.1 vscode:prepublish
> npm run compile


> teensysecurity@0.0.1 compile
> tsc -p ./

 ERROR  Command failed: npm list --production --parseable --depth=99999 --loglevel=error
npm ERR! code ELSPROBLEMS
npm ERR! invalid: vscode-arduino-api@0.2.1 /tmp/t/secure_plugin_vscode/node_modules/vscode-arduino-api

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/paul/.npm/_logs/2024-10-02T16_04_29_389Z-debug-0.log

```

Any chance you could point me to where I can get the stuff it wants to find in node_modules/vscode-arduino-api ?

**Reply by dankeboy36** [reply]
Great work. See my proposals at https://github.com/PaulStoffregen/secure_plugin_vscode/pull/1.

Using `webpack` is not mandatory but highly recommended.
